demos/text-generation/llm.js

```diff
@@ -247,75 +303,133 @@ export class LLM {
         // e.g. QWen2.0 supports max_length: 32768, in a matmul of the GQA decomposed op,
         // its input shapes will be [1, 14, 32768, 64] x [1, 14, 64, 32768] = [1, 14, 32768, 32768]
         // which exceeds the 2GB tensor size limitation.
-        const doPadding = this.provider === "webnn";
-        const inputIdsBuffer = doPadding ? this.paddingInput(inputIds, this.maxLength) : inputIds;
-        const positionIdsBuffer = doPadding ? this.paddingInput(positionIds, this.maxLength) : positionIds;
-
-        const attnMaskBuffer = this.paddingInput(attnMask, this.maxLength);
-        this.feed["input_ids"] = new ort.Tensor("int64", BigInt64Array.from(inputIdsBuffer), [
-            1,
-            inputIdsBuffer.length,
-        ]);
-        this.feed["attention_mask"] = new ort.Tensor("int64", BigInt64Array.from(attnMaskBuffer), [
-            1,
-            attnMaskBuffer.length,
-        ]);
-        this.feed["position_ids"] = new ort.Tensor("int64", BigInt64Array.from(positionIdsBuffer), [
-            1,
-            positionIdsBuffer.length,
-        ]);
+        if (this.provider == "webnn") {
+            inputIds = this.paddingInput(inputIds, this.maxLength);
+            positionIds = this.paddingInput(positionIds, this.maxLength);
+            attnMask = this.paddingInput(attnMask, this.maxLength);
+        }
+
+        this.feed["input_ids"] = new ort.Tensor("int64", BigInt64Array.from(inputIds), [1, inputIds.length]);
+        this.feed["attention_mask"] = new ort.Tensor("int64", BigInt64Array.from(attnMask), [1, attnMask.length]);
+        this.feed["position_ids"] = new ort.Tensor("int64", BigInt64Array.from(positionIds), [1, positionIds.length]);
         this.stop = false;
 
+        // shape of logits in prefill
+        const prefillLogitsShape = [1, this.provider == "webnn" ? this.maxLength : inputIdsLen, this.vocabSize];
+        const numElementsOfPrefillLogits = prefillLogitsShape.reduce((a, b) => a * b, 1);
         let lastToken = 0;
-        let outputs = await this.session1.run(this.feed);
-        lastToken = this.argmax(outputs["logits"], inputIdsLen);
-        let startLen = inputIdsLen;
+        if (this.provider == "webnn") {
+            this.fetches["logits"] = await createMlTensor(this.mlContext, "float16", prefillLogitsShape, false, true);
+        } else if (this.provider == "webgpu") {
+            const bufferSize = numElementsOfPrefillLogits * 2; // 2 bytes for float16
```

Comment:
The above code uses `TargetType.BYTES_PER_ELEMENT`. Can we use that here? I see `float16` repeated literally several times above - is there are parameter (for say float32 vs float16) somewhere above for that?